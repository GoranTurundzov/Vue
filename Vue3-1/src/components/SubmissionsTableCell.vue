<template>
    <div :class="direction">
        <div class="title text-capitalize">{{ title.replace('_', ' ') }}</div>
        <div class="font-w600 data-content" :class="title">{{ text }}</div>
    </div>
</template>

<script>
export default {
    name: 'SubmissionsTableRowColumn',
    props: {
        title: {
            required: true,
            description: 'The title of a single column from a table',
            type: String
        },
        text: {
            required: true,
            description: 'The content of a single column from a table',
            type: [String, Number, Boolean]
        },
        alignInRows: {
            type: Boolean,
            description: 'Column or Row that will affect how the cells will allign',
            default: false
        }
    },
    computed: {
        direction() {
            if (this.alignInRows == true) {
                return 'row border-b';
            }
            return 'column';
        }
    }
};
</script>
