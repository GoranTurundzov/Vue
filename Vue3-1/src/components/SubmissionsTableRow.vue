<template>
    <div class="content justify-center">
        <submissions-table-cell
            v-for="(key, value) in rowData"
            :title="value"
            :align-in-rows="alignInRows"
            :text="key"
            :key="key"
        ></submissions-table-cell>
        <div class="column" v-if="!alignInRows">
            <div>Action</div>
            <div>
                <button class="btn btn-gray btn-sm push-5-r" @click="editSubmission">
                    <i class="fa fa-edit"></i>
                </button>
                <button
                    class="btn btn-gray btn-sm push-5-r"
                    @click="activateModal"
                    data-testid="modal-activator"
                >
                    <i class="fa fa-eye"></i>
                </button>
                <button
                    class="btn btn-gray btn-sm push-5-r"
                    @click="deleteSubmission"
                    data-testid="delete-activator"
                >
                    <i class="fa fa-trash"></i>
                </button>
            </div>
        </div>
    </div>
</template>

<script>
import SubmissionsTableCell from './SubmissionsTableCell';

export default {
    components: {
        SubmissionsTableCell
    },
    name: 'SubmissionsTableRow',
    props: {
        rowData: {
            type: Object,
            required: true,
            description: 'A single row worth of data'
        },
        alignInRows: {
            type: Boolean,
            default: false,
            description: 'Direction on how the row will be printed (column =false; row = true)'
        }
    },
    emits: ['activate-modal', 'delete-submission', 'edit-submission'],
    methods: {
        activateModal() {
            this.$emit('activate-modal', this.rowData);
        },
        deleteSubmission() {
            this.$emit('delete-submission', this.rowData);
        },
        editSubmission() {
            this.$emit('edit-submission', this.rowData);
        }
    }
};
</script>

<style scoped>
.column {
    display: table-cell;
    text-align: left;
    min-width: 10.5vw;
    max-width: 10.5vw;
    font-size: 1.2rem;
    vertical-align: middle;
}
button:hover {
    transform: scale(1.2);
}
@media screen and (max-width: 1200px) {
    .content {
        flex-direction: column;
        padding: 5%;
        font-size: 1rem;
    }
    .column {
        display: block;
        width: 100%;
        min-width: 100%;
    }
}
</style>
