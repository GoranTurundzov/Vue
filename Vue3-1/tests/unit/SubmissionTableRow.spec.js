import { mount } from '@vue/test-utils';
import SubmissionsTableRow from '@/components/SubmissionsTableRow';
import SubmissionsTableCell from '@/components/SubmissionsTableCell';

let wrapper;

let spyActivateModalFn = jest.spyOn(SubmissionsTableRow.methods, 'activateModal');
let spyDeleteSubmissionFn = jest.spyOn(SubmissionsTableRow.methods, 'deleteSubmission');

describe('SubmissionsTableRow.vue', () => {
    const dataObject = {
        name: 'Jonathan',
        email: 'jonathan@gmail.com',
        recommendation: 'Maybe',
        feedback: 'My neighbor Karly has one of these. She works as a gambler and says it looks tall. I saw one of there in Sparly Islands and I bought One. works ok',
        date: '16/03/2021',
        form: 1,
        form_type: 'page',
        status: 'pending'
    };

    beforeEach(() => {
        wrapper = mount(SubmissionsTableRow, {
            propsData: {
                rowData: dataObject
            }
        });
    });

    it('Matches snapshot', () => expect(wrapper.element).toMatchSnapshot());

    test('It Renders the Table rows', () => {
        expect(wrapper.findComponent(SubmissionsTableCell).exists()).toBe(true);
    });

    it('Renders cells according to data length', () => {
        let cells = wrapper.findAllComponents(SubmissionsTableCell);
        expect(cells).toHaveLength(8);
    });

    test('Preview button emmits activate-modal', () => {
        wrapper.find('[data-testid="modal-activator"]').trigger('click');
        expect(spyActivateModalFn).toBeCalled();
        expect(wrapper.emitted('activate-modal')).toBeTruthy();
        expect(wrapper.emitted('activate-modal').length).toEqual(1);
        expect(wrapper.emitted('activate-modal')[0][0]).toEqual(dataObject);
    });

    test('Delete button (fa-trash) emits when clicked', () => {
        wrapper.find('[data-testid="delete-activator"]').trigger('click');
        expect(spyDeleteSubmissionFn).toBeCalled();
        expect(wrapper.emitted('delete-submission', wrapper.vm.rowData)).toBeTruthy();
        expect(wrapper.emitted('delete-submission')).toHaveLength(1);
        expect(wrapper.emitted('delete-submission')[0][0]).toEqual(dataObject);
    });
});
