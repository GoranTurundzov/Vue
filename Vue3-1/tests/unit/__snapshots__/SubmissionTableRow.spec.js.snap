// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionsTableRow.vue Matches snapshot 1`] = `
<div
  class="content justify-center"
>
  
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      name
    </div>
    <div
      class="font-w600 data-content name"
    >
      Jonathan
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      email
    </div>
    <div
      class="font-w600 data-content email"
    >
      jonathan@gmail.com
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      recommendation
    </div>
    <div
      class="font-w600 data-content recommendation"
    >
      Maybe
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      feedback
    </div>
    <div
      class="font-w600 data-content feedback"
    >
      My neighbor Karly has one of these. She works as a gambler and says it looks tall. I saw one of there in Sparly Islands and I bought One. works ok
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      date
    </div>
    <div
      class="font-w600 data-content date"
    >
      16/03/2021
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      form
    </div>
    <div
      class="font-w600 data-content form"
    >
      1
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      form type
    </div>
    <div
      class="font-w600 data-content form_type"
    >
      page
    </div>
  </div>
  <div
    class="column"
  >
    <div
      class="title text-capitalize"
    >
      status
    </div>
    <div
      class="font-w600 data-content status"
    >
      pending
    </div>
  </div>
  
  <div
    class="column"
  >
    <div>
      Action
    </div>
    <div>
      <button
        class="btn btn-gray btn-sm push-5-r"
      >
        <i
          class="fa fa-edit"
        />
      </button>
      <button
        class="btn btn-gray btn-sm push-5-r"
        data-testid="modal-activator"
      >
        <i
          class="fa fa-eye"
        />
      </button>
      <button
        class="btn btn-gray btn-sm push-5-r"
        data-testid="delete-activator"
      >
        <i
          class="fa fa-trash"
        />
      </button>
    </div>
  </div>
</div>
`;
